### DISCUSSION TITLE
Error when adding a Shape class

### DISCUSSION THREAD

**Discussion by RileyGe** [Macad3D/Macad3D#12] [Q&A] [answered]
I want to add a Vertex class to display a single point. I wrote the following class modeled on the Box class. However, 14 errors occurred in `SketchConstraintPointOnMidpoint`, `SketchConstraintPointOnSegment` and other classes.
The error info is: 
Cannot be converted from "ref Macad.SketchSolve.Point" to "ref Macad.Core.Shapes.Vertex"

![image](https://user-images.githubusercontent.com/26666371/124084115-3df47780-da81-11eb-9fb5-a608e74481ff.png)

```
using Macad.Common.Serialization;
using Macad.Occt;

namespace Macad.Core.Shapes
{
    [SerializeType]
    public sealed class Vertex : Shape
    {
        [SerializeMember]
        public double X
        {
            get
            {
                return _x;
            }
            set
            {
                if (_x != value)
                {
                    SaveUndo();
                    _x = value;
                    Invalidate();
                    RaisePropertyChanged();
                }
            }
        }
        [SerializeMember]
        public double Y
        {
            get
            {
                return _y;
            }
            set
            {
                if(_y != value)
                {
                    SaveUndo();
                    _y = value;
                    Invalidate();
                    RaisePropertyChanged();
                }
            }
        }
        [SerializeMember]
        public double Z
        {
            get
            {
                return _z;
            }
            set
            {
                if (_z != value)
                {
                    SaveUndo();
                    _z = value;
                    Invalidate();
                    RaisePropertyChanged();
                }
            }
        }

        private double _x;
        private double _y;
        private double _z;
        public override ShapeType ShapeType
        {
            get { return ShapeType.Solid; }
        }
        public static Vertex Create(double dimX, double dimY, double dimZ)
        {
            return new Vertex()
            {
                X = dimX,
                Y = dimY,
                Z = dimZ
            };
        }

        public Vertex()
        {
            _x = 0.0;
            _y = 0.0;
            _z = 0.0;
        }

        protected override bool MakeInternal(Shape.MakeFlags flags)
        {
            var makeVertex = new BRepBuilderAPI_MakeVertex(new Pnt(_x, _y, _z));
            BRep = makeVertex.Shape();
            return base.MakeInternal(flags);
        }
    }
}
```

**Comment by dipts** [comment, marked as answer]
Hi RileyGe,
the signature of this function is `public bool SetPoint(ref Point point, int pointKey, bool isConstant)`, did you changed that or created an overload?

What is the user story behind a vertex shape? Do you plan to build modifier/shaping tools on top of that? Then you need to define a new ShapeType, since "solid" should really contain valid TopoDS_Solid, as all modifier expect this type when feeded with a "solid" shape type. If the vertex is only for reference, you could better implement it as an "auxiliary", as recently introduced with the DatumPlane entity type.

**Reply by RileyGe** [reply]
Sorry, I changed the function `public bool SetPoint(ref Point point, int pointKey, bool isConstant)` by mistake!
I usually use measuring instruments to collect points, and then import them into CAD software for subsequent processing. Now I mainly use AutoCAD to process my data. But I want to use open source software to further automate my data processing process.
So adding and displaying points is very important to me, so I hope there is a direct tool to create and display points directly.
Yes, I need to display the vertices in the follow-up, and even need to modify the attributes of the points and measure the relative position between two or more points.
I will refer to `DatumPlane` to add my Vertex class, if there is any problem, I will trouble you again.

**Reply by RileyGe** [reply]
I read the `DatumPlane` and try to understand what the **auxiliary** present, I am pretty sure the Vertex is not auxiliary but something like geometry or shape. But what type of geometry or shape in Macad3D, I am not sure.
By the way, I think it's very strange we can not add a point in CAD software, do you consider to add this function in Macad3D?

**Reply by RileyGe** [reply]
The good news is that I have completed the function I want to complete, although this function is still very simple. I will gradually improve this function later.
I think I can write a simple tutorial on the development of this function.
This is what I have done: https://github.com/RileyGe/Macad3D/commit/e7bf4440b02a7f640ba475ec037c35d70e32ce66

**Reply by dipts** [reply]
> I read the `DatumPlane` and try to understand what the **auxiliary** present, I am pretty sure the Vertex is not auxiliary but something like geometry or shape. But what type of geometry or shape in Macad3D, I am not sure.
> By the way, I think it's very strange we can not add a point in CAD software, do you consider to add this function in Macad3D?

There is already enough free CAD software; these applications are typically defined as drawing applications. The goal of Macad, however, is to be a construction program for people who do not have a specialist training in technical drawing. The idea is to edit a single part (body) in its shape using modifiers. The basic primitive represents the starting point for the processing, currently either solid raw material (box, cyl, sphere) or a drawing (sketch). Maybe a wire in the future.

As I understand your user story, the dots are not used as primitives for further formation, but as a reference for measuring, positioning or aligning. This is what defines the newly created Auxiliary Group. These entities are defined independently, do not contribute anything to the shaping of a body, but serve as an aid to construction.
